src/dex/swaap-v1/swaap-v1-pool.ts

```diff
@@ -0,0 +1,897 @@
+import { DeepReadonly } from 'ts-essentials';
+import { BlockHeader } from '../../types';
+import { Address } from '../../types';
+import { IDexHelper } from '../../dex-helper/idex-helper';
+import {
+  SwaapV1PoolState,
+  SubgraphPoolBase,
+  SwaapV1PoolParameters,
+  SwaapV1PoolLiquidities,
+  SwaapV1PoolOracles,
+  OracleInitialData,
+  ChainLinkData,
+} from './types';
+import { StatefulEventSubscriber } from '../../stateful-event-subscriber';
+import { Log, MultiCallInput, MultiCallOutput } from '../../types';
+import PoolABI from '../../abi/swaap-v1/pool.json';
+import ProxyABI from '../../abi/chainlink.json';
+import Erc20ABI from '../../abi/erc20.json';
+import { Interface } from '@ethersproject/abi';
+
+export class SwaapV1Pool extends StatefulEventSubscriber<SwaapV1PoolState> {
+  static readonly poolInterface = new Interface(PoolABI);
+  static readonly proxyInterface = new Interface(ProxyABI);
+  static readonly erc20Interface = new Interface(Erc20ABI);
+  // LOG_CALL Event topics 0
+  static readonly SWAP_FEE_TOPIC =
+    '0x34e1990700000000000000000000000000000000000000000000000000000000';
+  static readonly LB_IN_ROUND_TOPIC =
+    '0x3cc3396000000000000000000000000000000000000000000000000000000000';
+  static readonly LB_STEP_IN_ROUND_TOPIC =
+    '0x045dbae000000000000000000000000000000000000000000000000000000000';
+  static readonly Z_TOPIC =
+    '0x3c515d9900000000000000000000000000000000000000000000000000000000';
+  static readonly HORIZON_TOPIC =
+    '0x314d204f00000000000000000000000000000000000000000000000000000000';
+  static readonly LB_IN_SEC_TOPIC =
+    '0x538e285c00000000000000000000000000000000000000000000000000000000';
+  static readonly MAX_UNPEG_TOPIC =
+    '0xacb21b3100000000000000000000000000000000000000000000000000000000';
+
+  static readonly ANSWER_UPDATED_TOPIC =
+    SwaapV1Pool.proxyInterface.getEventTopic('AnswerUpdated');
+  public addressSubscribers: Address[];
+  public oracleTokenMap: { [oracleAddress: Address]: Address };
+
+  constructor(
+    dexKey: string,
+    network: number,
+    protected pool: SubgraphPoolBase,
+    protected dexHelper: IDexHelper,
+  ) {
+    super(
+      `${dexKey}-${pool.id}`,
+      dexHelper.getLogger(`${dexKey}-${network}-${pool.id}`),
+    );
+
+    this.addressSubscribers = [pool.id];
+    this.oracleTokenMap = pool.tokens.reduce((acc, t) => {
+      this.addressSubscribers.push(
+        t.oracleInitialState.aggregator.toLowerCase(),
+      );
+      return {
+        ...acc,
+        [t.oracleInitialState.aggregator.toLowerCase()]: t.address,
+      };
+    }, {});
+  }
+
+  // Function which transforms the given state for the given log event.
+  // If the provided log does not affect the state, return null.
```

Comment:
If the provided log does not affect the state you should return previous state

Reply:
Returning `null` is actually correct if the state is not changed.  Doing so avoids unnecessarily creating an entry in the state history.

Reply:
ok perfect!